### ISSUE TITLE
Gradients == 0 (zero)

### ISSUE THREAD

**Issue by Roffild** [pytorch/pytorch#31396]
## 🐛 Bug

Python 3.7 / Pytorch 1.3.1 / Windows 10

Gradients on this data are not calculated. I can’t understand why.

```python
import numpy
import torch as th
from collections import OrderedDict

#dev = th.device("cuda")
dev = th.device("cpu")
dtype = th.float32

model = th.nn.Sequential(OrderedDict((
    ("input", th.nn.Linear(14650, 2)),
    ("output_activ", th.nn.Softmax(dim=-1))
)))

model.to(dev, dtype=dtype)
model.train()

optim = th.optim.Adam(model.parameters(), lr=1e-5)
loss = th.nn.BCELoss()

inputs = th.tensor(numpy.load("input.npy"), device=dev, dtype=dtype)
outputs = th.tensor(numpy.load("output.npy"), device=dev, dtype=dtype)

print("inputs = ", inputs)
print("outputs = ", outputs)

optim.zero_grad()
m = model(inputs)
lll = loss(m, outputs)
lll.backward()
print("Gradients:")
for p in model.parameters():
    if p.grad is None:
        continue
    grad = p.grad.data
    print(grad)
```
```
inputs =  tensor([ 5., 11., 11.,  ...,  0.,  1.,  0.])
outputs =  tensor([0., 1.])
Gradients:
tensor([[0., 0., 0.,  ..., 0., 0., 0.],
        [-0., -0., -0.,  ..., -0., -0., -0.]])
tensor([0., -0.])
```
[pytorch_grad.zip](https://github.com/pytorch/pytorch/files/3975975/pytorch_grad.zip)

**Comment by Enealor** [comment]
When running this myself, the gradients were zero because the variables `m` and `outputs` were equal. In this situation, it is correct for the gradients to be zero. To verify this reasoning, you can update the model parameters manually using
```
with th.no_grad():
    model.input.weight.fill_(1)
```
and test the model again. The second test will produce `m = [.5, .5]`, and the gradients will be non-zero for this model.

**Comment by albanD** [comment]
As mentioned by Enealor, the loss is `0` so the gradients are expected to be 0.

For further questions, please use the forum: https://discuss.pytorch.org/

**Comment by Roffild** [comment]
The initialization of the weights is always different on a new start, but the values of the gradients are always == 0.
```
inputs =  tensor([ 5., 11., 11.,  ...,  0.,  1.,  0.])
outputs =  tensor([0., 1.])
Result:  tensor([1., 0.], grad_fn=<SoftmaxBackward>)
Loss:  tensor(27.6310, grad_fn=<BinaryCrossEntropyBackward>)
Gradients:
tensor([[0., 0., 0.,  ..., 0., 0., 0.],
        [-0., -0., -0.,  ..., -0., -0., -0.]])
tensor([0., -0.])
```
```
inputs =  tensor([ 5., 11., 11.,  ...,  0.,  1.,  0.])
outputs =  tensor([0., 1.])
Result:  tensor([0., 1.], grad_fn=<SoftmaxBackward>)
Loss:  tensor(0., grad_fn=<BinaryCrossEntropyBackward>)
Gradients:
tensor([[0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]])
tensor([0., 0.])
```
If using other input values:
```python
inputs = th.randn(14650, device=dev, dtype=dtype)
```
```
inputs =  tensor([-1.0963,  2.2634,  0.4266,  ..., -0.7977, -1.0548,  1.2332])
outputs =  tensor([0., 1.])
Result:  tensor([0.6083, 0.3917], grad_fn=<SoftmaxBackward>)
Loss:  tensor(0.9372, grad_fn=<BinaryCrossEntropyBackward>)
Gradients:
tensor([[-0.6669,  1.3768,  0.2595,  ..., -0.4852, -0.6416,  0.7501],
        [ 0.6669, -1.3768, -0.2595,  ...,  0.4852,  0.6416, -0.7501]])
tensor([ 0.6083, -0.6083])
```

**Comment by albanD** [comment]
These look like the right gradients to me. Your result being [0,1] makes the gradients 0. But that is what they should be. Can you share the formula and value that you expect in this case?

**Comment by Roffild** [comment]
I agree that for the result [0, 1] gradients == 0.
But for the result [1, 0] Loss = 27.63, I expect gradients != 0. Looks like an overflow error.
```
inputs =  tensor([ 5., 11., 11.,  ...,  0.,  1.,  0.])
outputs =  tensor([0., 1.])
Result:  tensor([1., 0.], grad_fn=<SoftmaxBackward>)
Loss:  tensor(27.6310, grad_fn=<BinaryCrossEntropyBackward>)
Gradients:
tensor([[0., 0., 0.,  ..., 0., 0., 0.],
        [-0., -0., -0.,  ..., -0., -0., -0.]])
tensor([0., -0.])
```

**Comment by albanD** [comment]
But at the points where the softmax is saturated (the case here as your result is `tensor([1., 0.], grad_fn=<SoftmaxBackward>)`), it becomes a constant function and thus the gradients are 0 (up to floating point precision).
So if the forward values are `0.` and `1.`, I expect the gradient values to be `0.`.

**Comment by Enealor** [comment]
To add to what albanD said, I think the issue is partly a lack of transparency about how BCELoss is calculating the reported loss. When the model output is $x = [1, 0]$ and the desired output is $y = [0, 1]$, then the gradient is zero due to how the code is handling an edge case.

In particular, the binary cross-entropy between the two results should be infinite (due to the $-1\log(0)$ term), but the reported result is approximately 27. I couldn't find the implementation of the code (it's buried somewhere in the C++ code), but I'm guessing that a substitution for a constant is happening quietly and that this is the cause of the gradients being zero.
